File: android/src/main/java/im/zego/zego_express_engine/ZegoExpressEngineMediaPlayerEventHandler.java
//
//  ZegoExpressEngineMediaPlayerEventHandler.java
//  android
//  im.zego.zego_express_engine
//
//  Created by Patrick Fu on 2020/04/29.
//  Copyright © 2020 Zego. All rights reserved.
//

package im.zego.zego_express_engine;

import java.util.HashMap;

import im.zego.zegoexpress.ZegoMediaPlayer;
import im.zego.zegoexpress.callback.IZegoMediaPlayerEventHandler;
import im.zego.zegoexpress.constants.ZegoMediaPlayerNetworkEvent;
import im.zego.zegoexpress.constants.ZegoMediaPlayerState;
import io.flutter.plugin.common.EventChannel;

public class ZegoExpressEngineMediaPlayerEventHandler extends IZegoMediaPlayerEventHandler {

    private final EventChannel.EventSink sink;

    ZegoExpressEngineMediaPlayerEventHandler(EventChannel.EventSink sink) {
        super();
        this.sink = sink;
    }

    @Override
    public void onMediaPlayerStateUpdate(ZegoMediaPlayer mediaPlayer, ZegoMediaPlayerState state, int errorCode) {
        super.onMediaPlayerStateUpdate(mediaPlayer, state, errorCode);

        HashMap<String, Object> map = new HashMap<>();

        map.put("method", "onMediaPlayerStateUpdate");
        map.put("mediaPlayerIndex", mediaPlayer.getIndex());
        map.put("state", state.value());
        map.put("errorCode", errorCode);

        sink.success(map);
    }

    @Override
    public void onMediaPlayerNetworkEvent(ZegoMediaPlayer mediaPlayer, ZegoMediaPlayerNetworkEvent networkEvent) {
        super.onMediaPlayerNetworkEvent(mediaPlayer, networkEvent);

        HashMap<String, Object> map = new HashMap<>();

        map.put("method", "onMediaPlayerNetworkEvent");
        map.put("mediaPlayerIndex", mediaPlayer.getIndex());
        map.put("networkEvent", networkEvent.value());

        sink.success(map);
    }

    @Override
    public void onMediaPlayerPlayingProgress(ZegoMediaPlayer mediaPlayer, long millisecond) {
        super.onMediaPlayerPlayingProgress(mediaPlayer, millisecond);

        HashMap<String, Object> map = new HashMap<>();

        map.put("method", "onMediaPlayerPlayingProgress");
        map.put("mediaPlayerIndex", mediaPlayer.getIndex());
        map.put("millisecond", millisecond);

        sink.success(map);
    }
}

File: ios/Classes/ZegoExpressEngineMediaPlayerEventHandler.m
//
//  ZegoExpressEngineMediaPlayerEventHandler.m
//  Pods-Runner
//
//  Created by Patrick Fu on 2020/5/1.
//  Copyright © 2020 Zego. All rights reserved.
//

#import "ZegoExpressEngineMediaPlayerEventHandler.h"
#import "ZegoLog.h"

@interface ZegoExpressEngineMediaPlayerEventHandler ()

@property (nonatomic, strong) FlutterEventSink eventSink;

@end

@implementation ZegoExpressEngineMediaPlayerEventHandler

- (instancetype)initWithSink:(FlutterEventSink)sink {
    self = [super init];
    if (self) {
        _eventSink = sink;
    }
    return self;
}

#pragma mark - Zego

- (void)mediaPlayer:(ZegoMediaPlayer *)mediaPlayer stateUpdate:(ZegoMediaPlayerState)state errorCode:(int)errorCode {
    FlutterEventSink sink = _eventSink;
    ZGLog(@"state: %d, errorCode: %d", (int)state, errorCode);

    if (sink) {
        sink(@{
            @"method": @"onMediaPlayerStateUpdate",
            @"mediaPlayerIndex": mediaPlayer.index,
            @"state": @(state),
            @"errorCode": @(errorCode)
        });
    }
}

- (void)mediaPlayer:(ZegoMediaPlayer *)mediaPlayer networkEvent:(ZegoMediaPlayerNetworkEvent)networkEvent {
    FlutterEventSink sink = _eventSink;
    ZGLog(@"networkEvent: %d", (int)networkEvent);

    if (sink) {
        sink(@{
            @"method": @"onMediaPlayerNetworkEvent",
            @"mediaPlayerIndex": mediaPlayer.index,
            @"networkEvent": @(networkEvent)
        });
    }
}

- (void)mediaPlayer:(ZegoMediaPlayer *)mediaPlayer playingProgress:(unsigned long long)millisecond {
    FlutterEventSink sink = _eventSink;
    ZGLog(@"millisecond: %llu", millisecond);

    if (sink) {
        sink(@{
            @"method": @"onMediaPlayerPlayingProgress",
            @"mediaPlayerIndex": mediaPlayer.index,
            @"millisecond": @(millisecond)
        });
    }
}

@end
